lib/utils/makeValidator.js

```diff
@@ -3,21 +3,116 @@
+
+const processBaseError = (err, path) => {
+  const completePath = makePath(path, err.property).replace(
+    /\.instance\./g,
+    '.'
+  );
+
+  const subSchemas = err.message.match(/\[subschema \d+\]/g);
+  if (subSchemas) {
+    subSchemas.forEach((subschema, idx) => {
+      // err.schema is either an anonymous schema object or the name of a named schema
+      if (typeof err.schema === 'string') {
+        // this is basically only for FieldChoicesSchema and I'm not sure why
+        err.message += ' Consult the docs below for valid subschemas.';
+      } else {
+        // the subschemas have a type property
+        err.message = err.message.replace(
+          subschema,
+          err.schema[err.name][idx].type || 'unknown'
+        );
+      }
+    });
+  }
+
+  err.property = completePath;
+  return err;
+};
+
+/**
+ * We have a lot of `anyOf` schemas that return ambiguous errors. This recurses down the schema until it finds the errors that cause the failures replaces the ambiguity.
+ * @param {ValidationError} validationError an individual error
+ * @param {string} path current path in the error chain
+ * @param {Validator} validator validator object to pass around that has all the schemas
+ */
+const cleanError = (validationError, path, validator) => {
+  if (ambiguousTypes.includes(validationError.name)) {
+    // flatObjectSchema requires each property to be a type. instead of recursing down, it's more valuable to say "hey, it's not of these types"
+    if (validationError.argument.every(s => s.includes('subschema'))) {
+      return processBaseError(validationError, path);
+    }
+
+    // Try against each of A, B, and C to take a guess as to which it's closed to
+    // errorGroups will be an array of arrays of errors
+    const errorGroups = validationError.argument.map((schemaName, idx) => {
+      // this is what we'll validate against next
+      let nextSchema;
+      // schemaName is either "[subschema n]" or "/NamedSchema"
+      if (schemaName.startsWith('[subschema')) {
+        const maybeNamedSchema = validator.schemas[validationError.schema];
+
+        if (maybeNamedSchema) {
+          nextSchema = maybeNamedSchema[validationError.name][idx];
+        } else {
+          // hoist the anonymous subschema up
+          nextSchema = validationError.schema[validationError.name][idx];
+        }
+      } else {
+        nextSchema = validator.schemas[removeFirstAndLastChar(schemaName)];
+      }
+
+      const res = validator.validate(validationError.instance, nextSchema);
+      if (!res.errors.length) {
+        // This is a case that I can no longer reproduce, but it caused really weird behavior so I'm going to leave this as a canary for now
+        // in theory this should never happen - all schemas at this point should have errors
+        throw new Error(
+          'If you see this, please file an issue at https://github.com/zapier/zapier-platform-cli/issues'
+        );
+      }
```

Comment:
Throwing a hard error here blocks the dev from continuing their work, which is worse than the original behavior IMO. Maybe we want to take a milder approach here. Can we print a warning message and keep validation going, so the dev can see the validation errors at least?

I found one of the reasons why `res.errors` is an empty array below is because of this jsonschema [bug](https://github.com/tdegrunt/jsonschema/issues/263), where `validationError.instance` is undefined when the value being validated is falsy. To reproduce, try using a falsy value like an empty string in `inputFields`:

```js
{
  inputFields: [''],
}
```

To work around this bug, I suggest we do this before we do `validator.validate(validationError.instance, nextSchema)`:

```js
      if (validationError.instance === undefined) {
        // Work around a jsonschema bug: When the value being validated is a
        // falsy, such as false and '', validationError.instance isn't available.
        // See https://github.com/tdegrunt/jsonschema/issues/263
        const fullPath = path.replace(/^instance\./, '') + validationError.property.replace(/^instance\./, '.');
        validationError.instance = _.get(definition, fullPath);
      }
```

Important: You need to pass `definition` to `cleanErrors` if you want to use this code.

Reply:
ha, that's why I made [this change](https://github.com/zapier/zapier-platform-schema/pull/62/files#diff-f31cacec6592fa8d4d949b17e59b2669R49) but didn't connect the dots. Great find!

The reason I was throwing the error was that there _should_ be an error here (since validation has failed) but we have no error to throw. If we don't throw, we fail to validate but there's nothing to show, which I think is worse behavior. Given that, do you think it's appropriate to throw an error?

I do like pulling the instance out from above though, if we can.

Reply:
I was thinking like what if they have other errors? So for example, if I do `inputFields: [false]` and another violation like description is too short, this would just show me this block of red text:

![](https://zappy.zapier.com/DC8844F6-9619-4849-A8A7-3F36790D58E1.png)

I'd have no idea what's wrong with my code if I ever see this. However, if we print a warning message instead of throwing an error over this, we can still see the other error:

![](https://zappy.zapier.com/FB0B91A2-E516-40C9-8174-A5F09C5EA6C8.png)

All that say is even if the validation has failed on one branch, we can still keep running validation on the other branches. Does that make sense?

Reply:
ah, sure. I guess in my head we could return an error instead of throwing it because we don't want to actually halt execution. good call!